Fix W_y precision calculation

dmcdougall opened the pull request:
W_y is a precision matrix, not a covariance matrix.

Please merge #623 before merging this.  Also, this PR isn't complete.  There are *unit* (not regression) tests that fail, and I haven't figured out why yet.  I also need to add commits that:

1.  update the internal name of these matrices from the `*covariance*` flavour to the `*precision*` flavour;
2.  update the comments in the same manner.

dmcdougall committed:
Ensure GPMSA snarfs singular vectors in desc order

For the K matrix.

QUESO puts eigenvalues (with associated eigenvectors) in ascending
order.  GPMSA needs them in descending order.

This commit also updates the regression test solutions.

dmcdougall committed:
Fix Wy computation

Wy is a precision matrix, not a covariance matrix.  So it ought to be
filled with W_i on the diagonal, not W_i^{-1}.

This commit also updates the regression test solutions.

dmcdougall commented:
Oh I see why the unit tests are failing now.

roystgnr commented:
Didn't we decide that we need to rename m_observationErrorMatrix, now that W_y is actually holding precisions rather than errors?  I presume this has something to do with the scalar unit test breakage, because we actually are using that error matrix as an error matrix in that case?

dmcdougall commented:
Ok I've fixed the unit test failures.

I'd rather keep the m_emulatorErrorMatrix verbage around, because when we get to lnValue, the error matrices are better to work with than the precision matrices.  Check out the commit for the solution I implemented.

roystgnr commented on src/gp/src/GPMSA.C:
```diff
@@ -968,6 +972,9 @@ GPMSAFactory<V, M>::setUpEmulator()
 
+  // Emulator expects the error matrix, so invert Wy to get it.
+  m_observationErrorMatrix.reset(new M(Wy.inverse()));
+
```
This is a block diagonal matrix where the blocks are what the user provided.  Why are we first inverting the blocks, then making that big block diagonal matrix, then inverting the big matrix?

dmcdougall replied:
Because I'm lazy.  I'll update it to be less stupidly inefficient.

roystgnr replied:
Thanks!  The inefficiency doesn't bother me too too much (since it's once per run, not once per sample), but you had me questioning my sanity.  "Is there something wrong with my basic understanding of linear algebra identities?  Is the inverse-of-the-inverse somehow better for some numerical reason?"

dmcdougall replied:
No.  In fact the inverse-of-the-inverse is probably worse in terms of numerical accuracy?  ¯\\\_(ツ)\_/¯

dmcdougall replied:
Done.

roystgnr commented on src/gp/src/GPMSA.C:
```diff
@@ -894,25 +894,29 @@ GPMSAFactory<V, M>::setUpEmulator()
-  M& Wy = *m_observationErrorMatrix;
+
+  // Observation precision matrix
+  M Wy(env, Wy_row_map, Wyrows);
```
I guess it's okay for Wy to be a local matrix now because we only use it indirectly via BT_Wy_Binv?

dmcdougall replied:
Yep.

dmcdougall committed:
Use more efficient Sigma calculation

Instead of inverting a huge block diagonal matrix.

roystgnr commented on src/gp/src/GPMSA.C:
```diff
@@ -899,13 +899,17 @@ GPMSAFactory<V, M>::setUpEmulator()
     {
       const M & D_i = m_discrepancyMatrices[ex];
 
+      const M & Sigma_i = *m_observationErrorMatrices[ex];
```
You've got the handy readably-named reference here, why not use it when taking the inverse below?

roystgnr commented:
:+1: once Travis is happy.

dmcdougall committed:
Use more readable variable
